### ISSUE TITLE
Issue with subscription priorities

### ISSUE THREAD

**Issue by rwestphal** [sysrepo/sysrepo#1072]
Hi,

I want to subscribe to different parts of my YANG module and specify the order in which the configuration changes will be processed.

Here's a snippet of my YANG module:
```
  container ripd {
    container instance {
      presence "Present if the RIP protocol is enabled.";
      description
        "RIP routing instance.";

      leaf allow-ecmp {
        type boolean;
        default "false";
        description
          "Allow equal-cost multi-path.";
      }
      leaf default-metric {
        type uint8 {
          range "1..16";
        }
        default "1";
        description
          "Default metric of redistributed routes.";
      }
  }
```

And here's an example of how I'm subscribing to receive notifications about configuration changes:
```
sr_subtree_change_subscribe(session, "/frr-ripd:ripd/instance", ..., 100, ...);
sr_subtree_change_subscribe(session, "/frr-ripd:ripd/instance/default-metric", ..., 99, ...);
sr_subtree_change_subscribe(session, "/frr-ripd:ripd/instance/allow-ecmp", ..., 98, ...);
```

After changing the running configuration using _sysrepocfg_, my program is notified about the following changes:
```
/frr-ripd:ripd/instance
/frr-ripd:ripd/instance/allow-ecmp
/frr-ripd:ripd/instance/default-metric
/frr-ripd:ripd/instance/default-metric
/frr-ripd:ripd/instance/allow-ecmp
```

Notice how my program is being notified about `allow-ecmp` and `default-metric` twice. My assumption is that this happens because the subscription to `/frr-ripd:ripd/instance` encompasses all of its child elements. So subscribing to a container and its child elements leads to receiving duplicate notifications.

The biggest problem I see with this is that my program doesn't receive notifications in the intended order. In this example, my goal was to process the change in `default-metric` before the change in `allow-ecmp`, but this doesn't happen. Once the `instance` container is created, _sysrepo_ notifies about the creation of its child elements in the order they appear in the YANG module. Only after that that it triggers the subscriptions to the child elements, but at this point the subscription priorities are irrelevant because all configuration changes were processed already.

My understanding is that this is a bug. I think _sysrepo_ should trigger only the most specific subscription xpath for each configuration change. Does that make any sense?

Any help would be appreciated.

**Comment by michalvasko** [comment]
Hi,
it behaved exactly as I expected and now I also understand your output. You could have elaborated on how you got the output and saved me some time, but oh well.

So, I believe the lines
```
/frr-ripd:ripd/instance
/frr-ripd:ripd/instance/allow-ecmp
/frr-ripd:ripd/instance/default-metric
```
are from the first callback, 4th line from the second and 5th line from the third. You printed the changes you got from the session in the callback. Well, the changes are always recursive, that is how it works but there is little trouble in ignoring the changes you do not care about.

If you subscribed to the container and expect only changes concerning the container itself, loop through the changes until you find this one change and ignore the rest. Then everything will work exactly as you want it to.

Regards,
Michal

**Comment by rwestphal** [comment]
Hi Michal,

I apologize for not being more clear on that.

I tried your suggestion and it solved the problem for me. Since I'm using the same callback for everything, I decided to use the `private_ctx` argument of `sr_subtree_change_subscribe()` to specify which xpath the callback should process for that particular subscription. Not super optimal but it works. The only problem I have now is that ripd is taking a considerable amount of time to start because it's calling `sr_subtree_change_subscribe()` several times. I'll investigate this later and open an issue if necessary.

Thanks!
